### PR TITLE
Expose Cronet error message to the application layer.

### PR DESCRIPTION
**Pull request by yulin-liang**
Currently most of the error message users received are “UNAVAILABLE”, it doesn’t provide helpful information sometimes.So we are going to provide Cronet error message to the application layer.

### CODE REVIEW

File: src/core/ext/transport/cronet/transport/cronet_status.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+ *
+ * Copyright 2016 gRPC authors.
+ *
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.h:4]: 2020?

File: src/core/ext/transport/cronet/transport/cronet_status.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_EXT_TRANSPORT_CRONET_TRANSPORT_CRONET_STATUS_H
+#define GRPC_CORE_EXT_TRANSPORT_CRONET_TRANSPORT_CRONET_STATUS_H
+
+/*  HTTP/2 error codes are mapped to the following cronet net error codes */
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.h:22]: could you check if there is any similar implementation for Java?

File: src/core/ext/transport/cronet/transport/cronet_status.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+enum cronet_result {
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.h:23]: cronet_status_code or cronent_status_code_enum ?

File: src/core/ext/transport/cronet/transport/cronet_status.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  Cronet_RESULT_NULL_POINTER_CALLBACK = -306,
+  Cronet_RESULT_NULL_POINTER_EXECUTOR = -307,
+  Cronet_RESULT_NULL_POINTER_METHOD = -308,
+  Cronet_RESULT_NULL_POINTER_HEADER_NAME = -309,
+  Cronet_RESULT_NULL_POINTER_HEADER_VALUE = -310,
+  Cronet_RESULT_NULL_POINTER_PARAMS = -311,
+  Cronet_RESULT_NULL_POINTER_REQUEST_FINISHED_INFO_LISTENER_EXECUTOR = -312,
+};
+
+const char* cronet_result_to_string(cronet_result status);
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.h:57]: cronet_status_as_string() ?

File: src/core/ext/transport/cronet/transport/cronet_status.h (hunk 1 above)
Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.h:4]: Done, good catch!

File: src/core/ext/transport/cronet/transport/cronet_status.h (hunk 3 above)
Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.h]: Done.

File: src/core/ext/transport/cronet/transport/cronet_status.h (hunk 4 above)
Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.h]: Done.

File: src/core/ext/transport/cronet/transport/cronet_status.cc [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include "src/core/ext/transport/cronet/transport/cronet_status.h"
+
+const char* cronet_result_as_string(cronet_result_code status) {
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.cc:23]: cronet_status_as_string

File: src/core/ext/transport/cronet/transport/cronet_status.h (hunk 2 above)
Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.h:22]: gRPC-Java has the similar implementation. The `Cronet error` will be passed to the application layer as a part of `gRPC error`. Here's the code: https://github.com/grpc/grpc-java/blob/master/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java#L555-L561

File: src/core/ext/transport/cronet/transport/cronet_status.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  switch (status) {
+    case Cronet_RESULT_SUCCESS:
```
Comment (wenbozhu) [inline src/core/ext/transport/cronet/transport/cronet_status.cc:25]: s/cronet_RESULT/CRONET_STATUS/g

Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.cc]: Done.

File: src/core/ext/transport/cronet/transport/cronet_status.cc (hunk 5 above)
Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.cc]: Done.

File: src/core/ext/transport/cronet/transport/cronet_status.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+/*  HTTP/2 error codes are mapped to the following cronet net error codes */
+enum cronet_status_code {
+  CRONET_STATUS_SUCCESS = 0,
+  CRONET_STATUS_ILLEGAL_ARGUMENT = -100,
+  CRONET_STATUS_ILLEGAL_ARGUMENT_STORAGE_PATH_MUST_EXIST = -101,
+  CRONET_STATUS_ILLEGAL_ARGUMENT_INVALID_PIN = -102,
+  CRONET_STATUS_ILLEGAL_ARGUMENT_INVALID_HOSTNAME = -103,
+  CRONET_STATUS_ILLEGAL_ARGUMENT_INVALID_HTTP_METHOD = -104,
+  CRONET_STATUS_ILLEGAL_ARGUMENT_INVALID_HTTP_HEADER = -105,
+  CRONET_STATUS_ILLEGAL_STATE = -200,
```
Comment (zhungxd) [inline src/core/ext/transport/cronet/transport/cronet_status.h:31]: -106 INTERNET_DISCONNECTED is very useful

Comment (yulin-liang) [reply src/core/ext/transport/cronet/transport/cronet_status.h:31]: Looks like -106 is not in the [Cronet_RESULT enums](https://chromium.googlesource.com/chromium/src/+/master/components/cronet/native/generated/cronet.idl_c.h)(From line 71 to 103).

**Review by wenbozhu — APPROVED** [review]

**Review by yashykt — APPROVED** [review]

**Pull request merged**